src/utils/helpers.py

```diff
@@ -0,0 +1,95 @@
+import csv
+import itertools
+import json
+from argparse import Namespace
+from pathlib import Path
+
+
+def construct_predicates(
+    csv_lines: list[list[str]], taxyr: int | list[int] | None = None
+) -> list[str]:
+    """
+    Constructs SQL predicates based on the provided CSV lines and tax year(s).
+
+    Args:
+        csv_lines: A list of start and end PARIDs used to divide a table. The
+            third value is the rough number of PARIDs in the batch.
+        taxyr: A single tax year, a list of tax years, or None. If provided,
+            will be appended to the PARID predicates. Defaults to None.
+
+    Returns:
+        A list of SQL predicate strings used to divide a table into chunks.
+    """
+    taxyr_list = [taxyr] if isinstance(taxyr, int) else taxyr
+    if taxyr_list:
+        predicates = [
+            f"parid >= '{start}' AND parid <= '{end}' AND taxyr = {year}"
+            for (start, end, _), year in itertools.product(
+                csv_lines, taxyr_list
+            )
+        ]
+    else:
+        predicates = [
+            f"parid >= '{start}' AND parid <= '{end}'"
+            for start, end, _ in csv_lines
+        ]
+
+    return predicates
+
+
+def read_predicates(path: str) -> list[list[str]]:
+    """
+    Read a CSV file containing PARID start and end boundaries.
+
+    Args:
+        path: String path to the CSV file, relative to the Docker container.
+
+    Returns:
+        A list of start and end PARIDs used to divide a table. The third value
+        is the rough number of PARIDs in the batch.
+    """
+    pred_path = Path(path)
+    with open(pred_path.resolve().as_posix(), mode="r") as file:
+        csv_reader = csv.reader(file)
+        next(csv_reader)
+        csv_lines = [row for row in csv_reader]
+
+    return csv_lines
```

Comment:
**[Thought, non-blocking]** Since the predicate reading logic in this function seems to be a pretty simple operation, and its output only seems to be used by `construct_predicates`, I wonder if we could simplify things by just having `construct_predicates` accept a path to a predicates file and take responsibility for reading the predicates itself. I don't have strong feelings about it, just keeping an eye out for ways to keep the logic as simple as we can.

Reply:
I originally did this to avoid reading the CSV on every loop, but I reworked predicates entirely in fd7ffdc. Now this function expects a CSV path that can be specified per job which contains a column, start, and end value. These values are used to create arbitrary SQL BETWEEN expressions. I also removed the interaction with taxyr for simplicity.